doc/redact.1
.\" Automatically generated by Pandoc 3.0.1
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "REDACT" "1" "" "redact-haskell 0.6.0.0 (2026-01-01)" "redact Manual"
.nh
.SH NAME
.PP
\f[V]redact\f[R] - hide secret text on the terminal
.SH SYNOPSIS
.TP
\f[V]redact\f[R] [\f[I]COLOR_OPTIONS\f[R]] [\f[I]FILE_OPTION\f[R]]
redact input text using the configured color
.TP
\f[V]redact\f[R] [\f[I]COLOR_OPTIONS\f[R]] \f[V]--test\f[R]
redact test text using the configured color
.TP
\f[V]redact\f[R] \f[V]--colors\f[R]
redact test text using all possible colors
.SH DESCRIPTION
.PP
Redact is a utility for hiding secret text on the terminal.
.PP
It is designed to work with Markdown syntax.
Inline code (text enclosed in backticks) and code blocks that are fenced
with backticks are hidden.
Note that inline code that spans multiple lines is not supported, and
other types of code blocks are not hidden.
.PP
Text is hidden by displaying it using the same color for the foreground
and background.
When selecting hidden text, it remains hidden in some terminals and
becomes visible in other terminals.
.PP
Be sure to test it in your terminal before relying on it.
.SH OPTIONS
.TP
-h, --help
show help and exit
.TP
--version
show version and exit
.TP
-c, --color \f[I]COLOR\f[R]
redacted text color
.RS
.PP
The following colors are supported:
.IP \[bu] 2
\f[V]black\f[R]
.IP \[bu] 2
\f[V]red\f[R]
.IP \[bu] 2
\f[V]green\f[R]
.IP \[bu] 2
\f[V]yellow\f[R]
.IP \[bu] 2
\f[V]blue\f[R]
.IP \[bu] 2
\f[V]magenta\f[R]
.IP \[bu] 2
\f[V]cyan\f[R]
.IP \[bu] 2
\f[V]white\f[R]
.RE
.TP
-i, --intensity \f[I]INTENSITY\f[R]
redacted text color intensity
.RS
.PP
The following intensities are supported:
.IP \[bu] 2
\f[V]dull\f[R]
.IP \[bu] 2
\f[V]vivid\f[R]
.RE
.TP
-f, --file \f[I]PATH\f[R]
input file (default: \f[V]STDIN\f[R])
.TP
-l, --lenient
do not exit on parse errors
.RS
.PP
The following values are supported for the environment variable and
configuration:
.IP \[bu] 2
\f[V]true\f[R]
.IP \[bu] 2
\f[V]false\f[R]
.RE
.TP
--colors
redact test text using all possible colors
.TP
--test
redact test text using the configured color
.SH CONFIGURATION
.PP
Settings priority:
.IP "1." 3
command-line options
.RS 4
.IP \[bu] 2
\f[V]--color\f[R]
.IP \[bu] 2
\f[V]--intensity\f[R]
.IP \[bu] 2
\f[V]--lenient\f[R]
.RE
.IP "2." 3
environment variables
.RS 4
.IP \[bu] 2
\f[V]REDACT_COLOR\f[R]=\f[V]COLOR\f[R]
.IP \[bu] 2
\f[V]REDACT_INTENSITY\f[R]=\f[V]INTENSITY\f[R]
.IP \[bu] 2
\f[V]REDACT_LENIENT\f[R]=\f[V]LENIENT\f[R]
.RE
.IP "3." 3
settings file (\f[V]redact.ini\f[R])
.RS 4
.IP \[bu] 2
\f[V]color=COLOR\f[R]
.IP \[bu] 2
\f[V]intensity=INTENSITY\f[R]
.IP \[bu] 2
\f[V]lenient=LENIENT\f[R]
.RE
.IP "4." 3
defaults
.RS 4
.IP \[bu] 2
color: \f[V]red\f[R]
.IP \[bu] 2
intensity: \f[V]vivid\f[R]
.IP \[bu] 2
lenient: \f[V]false\f[R]
.RE
.SH EXIT CODES
.TP
0
no error
.TP
2
command-line error
.SH PROJECT
.TP
GitHub:
<https://github.com/ExtremaIS/redact-haskell>
.TP
Reporting issues:
GitHub: <https://github.com/ExtremaIS/redact-haskell/issues>
.RS
.PP
Email: <bugs@extrema.is>
.RE
.TP
Copyright
Copyright (c) 2020-2026 Travis Cardwell
.TP
License
The MIT License <https://opensource.org/licenses/MIT>


CHANGELOG.md
# `redact-haskell` Changelog

This project follows the [Haskell package versioning policy][PVP], with
versions in `A.B.C.D` format.  `A` may be incremented arbitrarily for
non-technical reasons, but [semantic versioning][SemVer] is otherwise
followed, where `A.B` is the major version, `C` is the minor version, and `D`
is the patch version.  Initial development uses versions `0.0.C.D`, for which
every version is considered breaking.

[PVP]: <https://pvp.haskell.org/>
[SemVer]: <https://semver.org/>

The format of this changelog is based on [Keep a Changelog][KaC], with the
following conventions:

* Level-two heading `Unreleased` is used to track changes that have not been
  released.
* Other level-two headings specify the release in `A.B.C.D (YYYY-MM-DD)`
  format, with newer versions above older versions.
* Level-three headings are used to categorize changes as follows:
    1. Breaking
    2. Non-Breaking
* Changes are listed in arbitrary order and present tense.

[KaC]: <https://keepachangelog.com/en/1.0.0/>

## Unreleased

### Non-Breaking

* Bump `base` dependency version upper bound
* Bump `optparse-applicative` dependency version upper bound
* Bump (vendored) `template-haskell` dependency version upper bound

## 0.6.0.0 (2024-12-30)

### Breaking

* Remove support for GHC 8.6, constraining lower bounds
* Remove support for GHC 8.4, constraining lower bounds
* Remove support for GHC 8.2, constraining lower bounds
* Change minimal Cabal from 1.24 to 3.0

### Non-Breaking

* Bump `ansi-terminal` dependency version upper bound
* Bump `base` dependency version upper bound
* Bump `tasty` dependency version upper bound
* Bump `text` dependency version upper bound
* Vendor `HMock` and `explainable-predicates`

## 0.5.0.0 (2023-05-28)

### Breaking

* Add support for `optparse-applicative` `0.18`

### Non-Breaking

* Bump `ansi-terminal` dependency version upper bound
* Bump `ansi-wl-pprint` dependency version upper bound
* Adjust dependency constraints to match tested versions
* Fix typo in CLI error message

## 0.4.0.0 (2022-03-02)

### Breaking

* Rewrite using a library
* Add `--lenient` flag, with strict behavior by default
* Reset colors on exceptions, including Control+C

### Non-Breaking

* Add tests
* Bump `optparse-applicative` dependency version upper bound

## 0.3.0.0 (2021-06-25)

### Breaking

* Fix `--help` when using `optparse-applicative` `0.16`

### Non-Breaking

* Refactor Nix configuration

## 0.2.0.1 (2021-05-27)

### Non-Breaking

* Fix link in README

## 0.2.0.0 (2021-05-27)

### Breaking

* Add support for `optparse-applicative` `0.16`

### Non-Breaking

* Add `.deb` and `.rpm` packaging
* Add Cabal support to `Makefile`
* Add Cabal tests to GitHub Actions
* Add [stan](https://hackage.haskell.org/package/stan) static analysis

## 0.1.0.0 (2020-11-28)

### Breaking

* Initial public release


redact.hs
------------------------------------------------------------------------------
-- |
-- Module      : Main
-- Description : redact: hide secret text on the terminal
-- Copyright   : Copyright (c) 2020-2022 Travis Cardwell
-- License     : MIT
--
-- See the README for details.
------------------------------------------------------------------------------

{-# LANGUAGE RecordWildCards #-}

module Main (main) where

-- https://hackage.haskell.org/package/ansi-terminal
import qualified System.Console.ANSI as Term

-- https://hackage.haskell.org/package/ansi-wl-pprint
import qualified Text.PrettyPrint.ANSI.Leijen as Doc
import Text.PrettyPrint.ANSI.Leijen (Doc)

-- https://hackage.haskell.org/package/base
import Control.Applicative ((<|>), optional)
import Control.Monad (foldM, forM_, unless)
import Data.Bifunctor (Bifunctor(bimap))
import Data.Char (isSpace, toLower)
import Data.Functor.Identity (Identity(runIdentity))
import Data.List (dropWhileEnd, intercalate)
import Data.Maybe (fromMaybe, isJust)
import Data.Version (showVersion)
import System.Environment (lookupEnv)
import System.Exit (ExitCode(ExitFailure), exitWith)
import System.IO
  ( Handle, IOMode(ReadMode), hIsEOF, hGetLine, hPutStrLn, stderr, stdin
  , withFile
  )

-- https://hackage.haskell.org/package/directory
import qualified System.Directory as Dir

-- https://hackage.haskell.org/package/optparse-applicative
import qualified Options.Applicative as OA

-- (redact:cabal)
import qualified Paths_redact as Project

-- (redact:executable)
import qualified LibOA

------------------------------------------------------------------------------
-- $Constants

defaultColor :: Term.Color
defaultColor = Term.Red

defaultIntensity :: Term.ColorIntensity
defaultIntensity = Term.Vivid

envColorName, envIntensityName :: String
envColorName     = "REDACT_COLOR"
envIntensityName = "REDACT_INTENSITY"

configFileName :: FilePath
configFileName = "redact.ini"

------------------------------------------------------------------------------
-- $Colors

colorMap :: [(String, Term.Color)]
colorMap =
    [ (toLower <$> show color, color)
    | color <- [minBound ..]
    ]

parseColor
  :: String
  -> Either String Term.Color
parseColor s
    = maybe (Left $ "uknown color: " ++ s) Right
    $ lookup (toLower <$> s) colorMap

------------------------------------------------------------------------------
-- $Intensities

intensityMap :: [(String, Term.ColorIntensity)]
intensityMap =
    [ (toLower <$> show intensity, intensity)
    | intensity <- [minBound ..]
    ]

parseIntensity
  :: String
  -> Either String Term.ColorIntensity
parseIntensity s
    = maybe (Left $ "unknown intensity: " ++ s) Right
    $ lookup (toLower <$> s) intensityMap

------------------------------------------------------------------------------
-- $Options

-- | Program options
data Options f
  = Options
    { optColor     :: !(f Term.Color)
    , optIntensity :: !(f Term.ColorIntensity)
    , optFile      :: !(Maybe FilePath)
    , optColors    :: !Bool
    , optTest      :: !Bool
    }

getOptions :: IO (Options Identity)
getOptions = do
    opts0 <- parseArgs

    opts1 <- if isJust (optColor opts0)
      then return opts0
      else do
        mColor <- getEnvColor
        return opts0 { optColor = mColor }
    opts2 <- if isJust (optIntensity opts1)
      then return opts1
      else do
        mIntensity <- getEnvIntensity
        return opts1 { optIntensity = mIntensity }

    case (optColor opts2, optIntensity opts2) of
      (Just color, Just intensity) -> return Options
        { optColor     = pure color
        , optIntensity = pure intensity
        , optFile      = optFile opts2
        , optColors    = optColors opts2
        , optTest      = optTest opts2
        }
      _otherOptions -> do
        configOpts <- loadConfigFile
        return Options
          { optColor = pure . fromMaybe defaultColor $
              optColor opts2 <|> configColor configOpts
          , optIntensity = pure . fromMaybe defaultIntensity $
              optIntensity opts2 <|> configIntensity configOpts
          , optFile   = optFile opts2
          , optColors = optColors opts2
          , optTest   = optTest opts2
          }

------------------------------------------------------------------------------
-- $Arguments

parseArgs :: IO (Options Maybe)
parseArgs = OA.execParser pinfo
  where
    pinfo :: OA.ParserInfo (Options Maybe)
    pinfo
      = OA.info (LibOA.helper <*> LibOA.versioner version <*> options)
      $ mconcat
          [ OA.fullDesc
          , OA.progDesc "hide secret text on the terminal"
          , OA.failureCode 2
          , OA.footerDoc $ Just footer
          ]

    version :: String
    version = "redact-haskell " ++ showVersion Project.version

    options :: OA.Parser (Options Maybe)
    options =
      Options
        <$> colorOption
        <*> intensityOption
        <*> fileOption
        <*> colorsOption
        <*> testOption

    footer :: Doc
    footer = colorsHelp LibOA.<||> intensitiesHelp LibOA.<||> settingsHelp

    colorsHelp :: Doc
    colorsHelp = LibOA.section "Colors:" . Doc.text $
      intercalate ", " (fst <$> colorMap)

    intensitiesHelp :: Doc
    intensitiesHelp = LibOA.section "Intensities:" . Doc.text $
      intercalate ", " (fst <$> intensityMap)

    settingsHelp :: Doc
    settingsHelp = LibOA.section "Settings priority:" . Doc.vcat $
      [ Doc.text "1. command-line options (--color, --intensity)"
      , Doc.text $
          "2. environment variables (" ++
          envColorName ++ ", " ++ envIntensityName ++ ")"
      , Doc.text ("3. settings file (" ++ configFileName ++ ")") Doc.<$$>
          ( Doc.indent 5 . Doc.vcat $
              [ Doc.text "color=COLOR"
              , Doc.text "intensity=INTENSITY"
              ]
          )
      , Doc.text "4. defaults" Doc.<$$>
          ( Doc.indent 5 . LibOA.table_ 2 $
              [ ["color:", toLower <$> show defaultColor]
              , ["intensity:", toLower <$> show defaultIntensity]
              ]
          )
      ]

colorOption :: OA.Parser (Maybe Term.Color)
colorOption = optional . OA.option (OA.eitherReader parseColor) $ mconcat
    [ OA.long "color"
    , OA.short 'c'
    , OA.metavar "COLOR"
    , OA.help "redacted text color"
    ]

intensityOption :: OA.Parser (Maybe Term.ColorIntensity)
intensityOption = optional . OA.option (OA.eitherReader parseIntensity) $
    mconcat
      [ OA.long "intensity"
      , OA.short 'i'
      , OA.metavar "INTENSITY"
      , OA.help "redacted text color intensity"
      ]

fileOption :: OA.Parser (Maybe FilePath)
fileOption = optional . OA.strOption $ mconcat
    [ OA.long "file"
    , OA.short 'f'
    , OA.metavar "PATH"
    , OA.help "input file (default: STDIN)"
    ]

colorsOption :: OA.Parser Bool
colorsOption = OA.switch $ mconcat
    [ OA.long "colors"
    , OA.help "redact test text using all possible colors"
    ]

testOption :: OA.Parser Bool
testOption = OA.switch $ mconcat
    [ OA.long "test"
    , OA.help "redact test text using the configured color"
    ]

------------------------------------------------------------------------------
-- $EnvVars

getEnv
  :: String
  -> (String -> Either String a)
  -> IO (Maybe a)
getEnv envName parse = do
    meex <- fmap parse <$> lookupEnv envName
    case meex of
      Nothing -> return Nothing
      Just (Right x) -> return $ Just x
      Just (Left err) -> errorExit $
        "environment variable " ++ envName ++ ": " ++ err

getEnvColor :: IO (Maybe Term.Color)
getEnvColor = getEnv envColorName parseColor

getEnvIntensity :: IO (Maybe Term.ColorIntensity)
getEnvIntensity = getEnv envIntensityName parseIntensity

------------------------------------------------------------------------------
-- $ConfigFile

data ConfigFileOptions
  = ConfigFileOptions
    { configColor     :: !(Maybe Term.Color)
    , configIntensity :: !(Maybe Term.ColorIntensity)
    }

loadConfigFile :: IO ConfigFileOptions
loadConfigFile = do
    path <- Dir.getXdgDirectory Dir.XdgConfig configFileName
    exists <- Dir.doesFileExist path
    if exists
      then foldM go defOpts . zip [1..] . map strip . lines =<< readFile path
      else return defOpts
  where
    defOpts :: ConfigFileOptions
    defOpts = ConfigFileOptions
      { configColor     = Nothing
      , configIntensity = Nothing
      }

    go :: ConfigFileOptions -> (Int, String) -> IO ConfigFileOptions
    go opts (_, "") = return opts
    go opts (_, '#' : _) = return opts
    go opts (lineNum, line) = case bimap strip strip (break (== '=') line) of
      ("", _) -> errorExit $ "config file: empty key on line " ++ show lineNum
      (key, '=' : value) -> case key of
        "color" -> case parseColor value of
          Right color -> return opts { configColor = Just color }
          Left err -> errorExit $
            "config file: color error on line " ++ show lineNum ++ ": " ++ err
        "intensity" -> case parseIntensity value of
          Right intensity -> return opts { configIntensity = Just intensity }
          Left err -> errorExit $
            "config file: intensity error on line " ++ show lineNum ++ ": " ++
            err
        _otherKey -> return opts
      _mistmatch ->
        errorExit $ "config file: unable to parse line " ++ show lineNum

strip :: String -> String
strip = dropWhile isSpace . dropWhileEnd isSpace
{-# INLINE strip #-}

------------------------------------------------------------------------------
-- $Implementation

redactLine :: Term.Color -> Term.ColorIntensity -> String -> IO ()
redactLine color intensity = go
  where
    go :: String -> IO ()
    go "" = putStrLn ""
    go ('`' : s) = do
      putStr "`"
      let (hiddenText, rest) = break (== '`') s
      unless (null hiddenText) $ do
        Term.setSGR
          [ Term.SetColor Term.Foreground intensity color
          , Term.SetColor Term.Background intensity color
          ]
        putStr hiddenText
        Term.setSGR [Term.Reset]
      case rest of
        '`' : s' -> putStr "`" >> go s'
        _emptyString -> putStrLn ""
    go s = do
      let (clearText, rest) = break (== '`') s
      putStr clearText
      go rest

runHandle :: Term.Color -> Term.ColorIntensity -> Handle -> IO ()
runHandle color intensity handle = go =<< hIsEOF handle
  where
    go :: Bool -> IO ()
    go True  = return ()
    go False = do
      s <- hGetLine handle
      if getFenceLength s >= 3
        then do
          putStrLn s
          Term.setSGR
            [ Term.SetColor Term.Foreground intensity color
            , Term.SetColor Term.Background intensity color
            ]
          goFence s =<< hIsEOF handle
        else do
          redactLine color intensity s
          go =<< hIsEOF handle

    getFenceLength :: String -> Int
    getFenceLength s
      | all (== '`') s = length s
      | otherwise      = 0

    goFence :: String -> Bool -> IO ()
    goFence _     True  = Term.setSGR [Term.Reset]
    goFence fence False = do
      s <- hGetLine handle
      if s == fence
        then do
          Term.setSGR [Term.Reset]
          putStrLn s
          go =<< hIsEOF handle
        else do
          putStrLn s
          goFence fence =<< hIsEOF handle

runFile :: Term.Color -> Term.ColorIntensity -> FilePath -> IO ()
runFile color intensity path =
    withFile path ReadMode $ runHandle color intensity

runTest :: Term.Color -> Term.ColorIntensity -> IO ()
runTest color intensity = redactLine color intensity $
    (toLower <$> show color) ++ " " ++ (toLower <$> show intensity) ++
    ": `hidden`"

runColors :: IO ()
runColors =
    forM_ [minBound ..] $ \color ->
      forM_ [minBound ..] $ \intensity ->
        runTest color intensity

------------------------------------------------------------------------------
-- $CLI

errorExit :: String -> IO a
errorExit message = do
    hPutStrLn stderr $ "error: " ++ message
    exitWith $ ExitFailure 1

main :: IO ()
main = do
    Options{..} <- getOptions
    let color     = runIdentity optColor
        intensity = runIdentity optIntensity
    case (optColors, optTest, optFile) of
      (True, _,    _)         -> runColors
      (_,    True, _)         -> runTest color intensity
      (_,    _,    Nothing)   -> runHandle color intensity stdin
      (_,    _,    Just path) -> runFile color intensity path
